src/supplemental_attributes.jl

```diff
@@ -1,76 +1,109 @@
+    associations::SupplementalAttributeAssociations
 end
 
-get_member_string(::SupplementalAttributes) = "supplemental attributes"
+function SupplementalAttributeManager(
+    data::SupplementalAttributesByType,
+    time_series_manager::TimeSeriesManager,
+)
+    return SupplementalAttributeManager(
+        data,
+        time_series_manager,
+        SupplementalAttributeAssociations(get_database(time_series_manager), true),
+    )
+end
 
-function SupplementalAttributes(time_series_manager::TimeSeriesManager)
-    return SupplementalAttributes(SupplementalAttributesContainer(), time_series_manager)
+function SupplementalAttributeManager(time_series_manager::TimeSeriesManager)
+    return SupplementalAttributeManager(
+        SupplementalAttributesByType(),
+        time_series_manager,
+        SupplementalAttributeAssociations(get_database(time_series_manager), true),
+    )
 end
 
+get_member_string(::SupplementalAttributeManager) = "supplemental attributes"
+
 function add_supplemental_attribute!(
-    supplemental_attributes::SupplementalAttributes,
+    mgr::SupplementalAttributeManager,
     component::InfrastructureSystemsComponent,
-    supplemental_attribute::SupplementalAttribute;
-    kwargs...,
+    attribute::SupplementalAttribute;
+    allow_existing_time_series = false,
 )
-    try
-        attach_component!(supplemental_attribute, component)
-        attach_supplemental_attribute!(component, supplemental_attribute)
-        _add_supplemental_attribute!(
-            supplemental_attributes,
-            supplemental_attribute;
-            kwargs...,
+    if has_association(mgr.associations, component, attribute)
+        throw(
+            ArgumentError(
+                "There is already an association between " *
+                "$(summary(component)) and $(summary(attribute))",
+            ),
         )
-    catch e
-        detach_component!(supplemental_attribute, component)
-        detach_supplemental_attribute!(component, supplemental_attribute)
-        rethrow(e)
     end
+    _attach_attribute!(
+        mgr,
+        attribute;
+        allow_existing_time_series = allow_existing_time_series,
+    )
+    add_association!(mgr.associations, component, attribute)
     return
 end
 
-function _add_supplemental_attribute!(
-    supplemental_attributes::SupplementalAttributes,
-    supplemental_attribute::T;
+function _attach_attribute!(
+    mgr::SupplementalAttributeManager,
+    attribute::SupplementalAttribute;
     allow_existing_time_series = false,
-) where {T <: SupplementalAttribute}
-    if !isconcretetype(T)
-        throw(ArgumentError("add_supplemental_attribute! only accepts concrete types"))
-    end
+)
+    is_attached(attribute, mgr) && return
 
-    supplemental_attribute_uuid = get_uuid(supplemental_attribute)
-    if isempty(get_component_uuids(supplemental_attribute))
+    if !allow_existing_time_series && has_time_series(attribute)
         throw(
             ArgumentError(
```

Comment:
Should we still show the supplemental attribute type in the error?

Reply:
Yeah. I'll change this to call `summary` on the attribute.